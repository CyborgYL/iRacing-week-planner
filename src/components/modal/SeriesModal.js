// @flow

import * as React from 'react';
import { useTranslation } from 'react-i18next';
import moment from 'moment';
import classnames from 'classnames';
import Modal from './Modal';
import RaceLength from '../columns/RaceLength';

import allRaces from '../../lib/races';

import styles from '../../styles/main.module.scss';

type Props = {
  onClose: () => void,
  isOpen: boolean,
  ownedTracks: Array<number>,
  seriesId: number,
};

export default function SeriesModal({ onClose, ownedTracks, isOpen, seriesId }: Props) {
  const races = allRaces.filter((race) => race.seriesId === seriesId);
  const { t } = useTranslation();

  const now = moment().utc();

  return (
    <Modal
      isOpen={isOpen}
      onClose={onClose}
      title={t('Tracks for {{series}}', {
        series: t(races[0].series),
      })}
      doneAction={onClose}
    >
      <div className={styles['container-fluid']}>
        <div className={styles['table-responsive']}>
          <table className={styles.table} style={{ fontSize: '0.8em' }}>
            <thead>
              <tr>
                <th>{t('Week')}</th>
                <th>{t('Track')}</th>
                <th>{t('Start')}</th>
                <th>{t('End')}</th>
                <th>{t('Length')}</th>
              </tr>
            </thead>
            <tbody>
              {races.map((race) => {
                const raceWeekEnd = moment(race.startTime).add(race.weekLength);
                const current = now.isBetween(race.startTime, raceWeekEnd);

                const startDate = moment(race.startTime).local().toDate();
                const endDate = moment(race.startTime).local().add(race.weekLength)
                  .subtract(1, 'days')
                  .toDate();

                return (
                  <tr key={race.week} style={current ? { fontWeight: 700 } : {}}>
                    <td>
                      {race.week + 1}
                    </td>
                    <td className={classnames({ [styles.success]: ownedTracks.indexOf(race.trackId) !== -1 })}>
                      {t(race.track)}
                    </td>
                    <td>
                      {t('{{date, YYYY-MM-DD}}', { date: startDate })}
                    </td>
                    <td>
                      {t('{{date, YYYY-MM-DD}}', { date: endDate })}
                    </td>
                    <RaceLength race={race} />
                  </tr>
                );
              })}
            </tbody>
          </table>
        </div>
      </div>
    </Modal>
  );
}
