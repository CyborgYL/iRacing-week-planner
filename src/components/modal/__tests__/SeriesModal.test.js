import { describe, test } from '@jest/globals';
import React from 'react';
import renderer, { act } from 'react-test-renderer';
import MockDate from 'mockdate';
import SeriesModal from '../SeriesModal';

jest.mock('firebase/auth');

describe('components/modal/SeriesModal', () => {
  test('renders closed', async () => {
    const onClose = jest.fn(() => {});
    const component = renderer.create(
      <SeriesModal isOpen={false} onClose={onClose} ownedTracks={[]} seriesId={245} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('renders open', async () => {
    MockDate.set('2022-09-07T13:30:00.000Z');
    const onClose = jest.fn(() => {});
    let component;
    act(() => {
      component = renderer.create(
        <SeriesModal isOpen onClose={onClose} ownedTracks={[6]} seriesId={245} />,
      );
    });

    expect(component.toJSON()).toMatchSnapshot();

    await act(async () => {
      await component.root.findByProps({ 'aria-label': 'Close' }).props.onClick({
        preventDefault: () => {},
        stopPropagation: () => {},
      });
    });

    expect(onClose).toHaveBeenCalled();
  });
});
