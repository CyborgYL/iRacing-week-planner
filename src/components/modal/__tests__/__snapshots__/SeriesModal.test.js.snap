// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/modal/SeriesModal renders closed 1`] = `null`;

exports[`components/modal/SeriesModal renders open 1`] = `
<div
  className="ReactModal__Overlay"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(255, 255, 255, 0.75)",
      "bottom": 0,
      "left": 0,
      "position": "fixed",
      "right": 0,
      "top": 0,
    }
  }
>
  <div
    aria-modal={true}
    className="ReactModal__Content Modal__Bootstrap modal-dialog modal-lg"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    role="dialog"
    style={Object {}}
    tabIndex="-1"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <button
          aria-label="Close"
          className="close"
          onClick={[Function]}
          type="button"
        >
          <span
            className="glyphicon glyphicon-remove"
          />
        </button>
        <h4
          className="modal-title"
        >
          Tracks for Fanatec Global Mazda MX-5 Cup
        </h4>
      </div>
      <div
        className="modal-body"
        style={
          Object {
            "maxHeight": "55vh",
            "overflowY": "auto",
          }
        }
      >
        <div
          className="container-fluid"
        >
          <div
            className="table-responsive"
          >
            <table
              className="table"
              style={
                Object {
                  "fontSize": "0.8em",
                }
              }
            >
              <thead>
                <tr>
                  <th>
                    Week
                  </th>
                  <th>
                    Track
                  </th>
                  <th>
                    Start
                  </th>
                  <th>
                    End
                  </th>
                  <th>
                    Length
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr
                  style={Object {}}
                >
                  <td>
                    1
                  </td>
                  <td
                    className=""
                  >
                    Tsukuba Circuit - 2000 Full
                  </td>
                  <td>
                    2020-09-15
                  </td>
                  <td>
                    2020-09-21
                  </td>
                  <td>
                    <div>
                      15L
                    </div>
                  </td>
                </tr>
                <tr
                  style={Object {}}
                >
                  <td>
                    2
                  </td>
                  <td
                    className=""
                  >
                    Okayama International Circuit - Full Course
                  </td>
                  <td>
                    2020-09-22
                  </td>
                  <td>
                    2020-09-28
                  </td>
                  <td>
                    <div>
                      25L
                    </div>
                  </td>
                </tr>
                <tr
                  style={Object {}}
                >
                  <td>
                    3
                  </td>
                  <td
                    className="success"
                  >
                    Lime Rock Park - Grand Prix
                  </td>
                  <td>
                    2020-09-29
                  </td>
                  <td>
                    2020-10-05
                  </td>
                  <td>
                    <div>
                      15L
                    </div>
                  </td>
                </tr>
                <tr
                  style={
                    Object {
                      "fontWeight": 700,
                    }
                  }
                >
                  <td>
                    4
                  </td>
                  <td
                    className="success"
                  >
                    Summit Point Raceway - Summit Point Raceway
                  </td>
                  <td>
                    2020-10-06
                  </td>
                  <td>
                    2020-10-12
                  </td>
                  <td>
                    <div>
                      15L
                    </div>
                  </td>
                </tr>
                <tr
                  style={Object {}}
                >
                  <td>
                    5
                  </td>
                  <td
                    className=""
                  >
                    Oulton Park Circuit - International
                  </td>
                  <td>
                    2020-10-13
                  </td>
                  <td>
                    2020-10-19
                  </td>
                  <td>
                    <div>
                      15L
                    </div>
                  </td>
                </tr>
                <tr
                  style={Object {}}
                >
                  <td>
                    6
                  </td>
                  <td
                    className=""
                  >
                    Charlotte Motor Speedway - Roval
                  </td>
                  <td>
                    2020-10-20
                  </td>
                  <td>
                    2020-10-26
                  </td>
                  <td>
                    <div>
                      20L
                    </div>
                  </td>
                </tr>
                <tr
                  style={Object {}}
                >
                  <td>
                    7
                  </td>
                  <td
                    className=""
                  >
                    Oran Park Raceway - Grand Prix
                  </td>
                  <td>
                    2020-10-27
                  </td>
                  <td>
                    2020-11-02
                  </td>
                  <td>
                    <div>
                      15L
                    </div>
                  </td>
                </tr>
                <tr
                  style={Object {}}
                >
                  <td>
                    8
                  </td>
                  <td
                    className=""
                  >
                    Okayama International Circuit - Short
                  </td>
                  <td>
                    2020-11-03
                  </td>
                  <td>
                    2020-11-09
                  </td>
                  <td>
                    <div>
                      15L
                    </div>
                  </td>
                </tr>
                <tr
                  style={Object {}}
                >
                  <td>
                    9
                  </td>
                  <td
                    className="success"
                  >
                    Summit Point Raceway - Jefferson Circuit
                  </td>
                  <td>
                    2020-11-10
                  </td>
                  <td>
                    2020-11-16
                  </td>
                  <td>
                    <div>
                      15L
                    </div>
                  </td>
                </tr>
                <tr
                  style={Object {}}
                >
                  <td>
                    10
                  </td>
                  <td
                    className=""
                  >
                    Oulton Park Circuit - Fosters
                  </td>
                  <td>
                    2020-11-17
                  </td>
                  <td>
                    2020-11-23
                  </td>
                  <td>
                    <div>
                      25L
                    </div>
                  </td>
                </tr>
                <tr
                  style={Object {}}
                >
                  <td>
                    11
                  </td>
                  <td
                    className=""
                  >
                    Oran Park Raceway - South
                  </td>
                  <td>
                    2020-11-24
                  </td>
                  <td>
                    2020-11-30
                  </td>
                  <td>
                    <div>
                      20L
                    </div>
                  </td>
                </tr>
                <tr
                  style={Object {}}
                >
                  <td>
                    12
                  </td>
                  <td
                    className=""
                  >
                    WeatherTech Raceway at Laguna Seca - Full Course
                  </td>
                  <td>
                    2020-12-01
                  </td>
                  <td>
                    2020-12-07
                  </td>
                  <td>
                    <div>
                      15L
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
      <div
        className="modal-footer"
      >
        <button
          className="btn btn-primary"
          onClick={[Function]}
          type="button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;
